api-metastore/src/main/java/org/zalando/nakadi/controller/PartitionsController.java

```diff
@@ -103,59 +111,174 @@ private static NakadiCursor selectLast(final List<Timeline> activeTimelines, fin
                     first.getPartition(),
                     cursorConverter.convert(first.getFirst()).getOffset(),
                     cursorConverter.convert(selectLast(timelines, last, first)).getOffset());
         }).collect(Collectors.toList());
-        return ok().body(result);
+    }
+
+    private List<Cursor> getParsedCursors(final String cursorsString, final String eventType){
+        if(StringUtils.isEmpty(cursorsString)){
+            return emptyList();
+        }
+        try {
+            final List<Cursor> cursorList = objectMapper.
+                    readValue(
+                        cursorsString,
+                        new TypeReference<List<Cursor>>() {
+                    });
+
+            validateCursors(cursorList, eventType);
+            return cursorList;
+        } catch (JsonProcessingException e) {
+            throw new UnparseableCursorException("malformed cursors", e, cursorsString);
+        }
+
+    }
+
+    private void validateCursors(final List<Cursor> cursorList, final String eventType) {
+        for (final Cursor cursor : cursorList) {
+            if (StringUtils.isEmpty(cursor.getPartition())) {
+                throw new InvalidCursorException(CursorError.NULL_PARTITION, cursor, eventType);
+            } else if (StringUtils.isEmpty(cursor.getOffset())) {
+                throw new InvalidCursorException(CursorError.NULL_OFFSET, cursor, eventType);
+            }
+        }
+    }
+
+    private List<PartitionStatistics> initializeLastStats( final List<Timeline> timelines,
+                                                           final List<PartitionStatistics> firstStats) {
+        final List<PartitionStatistics> lastStats;
+        if (timelines.size() == 1) {
+            lastStats = firstStats;
+        } else {
+            lastStats = timelineService.
+                    getTopicRepository(timelines.get(timelines.size() - 1))
+                    .loadTopicStatistics(singletonList(timelines.get(timelines.size() - 1)));
+        }
+        return lastStats;
+    }
+
+    private List<EventTypePartitionView> getViewsWithUnconsumedEvents
+            (final String eventTypeName,
+             final List<Cursor> cursorList) {
+
+        final List<EventTypePartitionView> cursorLags
+                = cursorOperationsService
+                .toCursorLagStream(
+                        cursorList,
+                        eventTypeName,
+                        cursorConverter)
+                .map(cl -> new EventTypePartitionView(
+                                cl.getPartition(),
+                                cl.getOldestAvailableOffset(),
+                                cl.getNewestAvailableOffset(),
+                                cl.getUnconsumedEvents())
+                ).collect(Collectors.toList());
+
+        if (cursorLags.size() != cursorList.size()){
+            throw new NakadiBaseException(String.format("Problem with cursors '%s'", cursorLags));
```

Comment:
Can we provide a more specific message than just "problem with cursors"?

Reply:
Sure, earlier there was just an empty exception being thrown. Will come up with better message.

Reply:
done